browser/fxr/content/permissions.js

```diff
@@ -42,65 +50,94 @@ class FxrPermissionPromptPrototype {
       this.deny();
     }
 
-    this.responseCallback();
+    if (this.ui) {
+      this.ui.close();
+      this.ui = null;
+    }
+
+    let status = {};
+    status[allowed ? "allowed" : "denied"] = this.promptTypes;
+    this.responseCallback(status);
   }
 }
 
 // WebRTC-specific class implementation
+//
+// For now-- only supports microphone permissions with the following behavior:
+// - per domain/principal, allowed is required for every request
+// - per domain/principal, denied is persisted for the session (i.e., must
+//   restart the browser )
+// The UI behavior to support this also includes code in WebRTCParent.jsm and
+// fxrui.js.
 class FxrWebRTCPrompt extends FxrPermissionPromptPrototype {
-  showPrompt() {
-    for (let typeName of this.request.requestTypes) {
-      if (typeName !== "Microphone" && typeName !== "Camera") {
-        // Only Microphone and Camera requests are allowed. Automatically deny
-        // any other request.
-        this.defaultDeny();
-        return;
-      }
-    }
-
-    super.showPrompt();
-  }
+  constructor(aRequest, aBrowser, aCallback) {
+    super(aRequest, aBrowser, aCallback);
 
-  allow() {
     let { audioDevices, videoDevices } = this.request;
 
-    let principal = Services.scriptSecurityManager.createContentPrincipalFromOrigin(
+    this.principal = Services.scriptSecurityManager.createContentPrincipalFromOrigin(
```

Comment:
Can you use the content principal of the browser instead?
Constructing new principals should be avoided if possible.

Reply:
👍